### PR TITLE
scalar/coefficient types in Polyhedral Geometry

### PR DESCRIPTION
**Pull request by alexej-jordan**
Current state: all objects within Polyhedral Geometry can be templated, but only with `fmpq`. existing tests and doctests are adjusted and running.

Next important steps: support number field coefficients (degree 2) with polymake's `QuadraticExtension`; add tests for different scalars and for `PolyhedralComplex`.

### CODE REVIEW

File: src/PolyhedralGeometry/Cone/constructors.jl [hunk 1]
```diff
@@ -57,7 +57,7 @@ end
 
 
 @doc Markdown.doc"""
-    positive_hull(R::Union{Oscar.MatElem, AbstractMatrix, SubObjectIterator})
+    positive_hull([:.Type{T} = fmpq,] R::Union{Oscar.MatElem, AbstractMatrix, SubObjectIterator})
```
Comment (benlorenz) [inline src/PolyhedralGeometry/Cone/constructors.jl]:
```suggestion
    positive_hull([::Type{T} = fmpq,] R::Union{Oscar.MatElem, AbstractMatrix, SubObjectIterator})
```

File: src/PolyhedralGeometry/Cone/properties.jl [hunk 2]
```diff
@@ -198,7 +198,7 @@ julia> f_vector(square)
 function f_vector(C::Cone)
     pmc = pm_object(C)
     ldim = pmc.LINEALITY_DIM
-    return vcat(fill(0,ldim),pmc.F_VECTOR)
+    return Vector{Int}(vcat(fill(0, ldim), pmc.F_VECTOR))
```
Comment (benlorenz) [inline src/PolyhedralGeometry/Cone/properties.jl]:
```suggestion
    return Vector{fmpz}(vcat(fill(0, ldim), pmc.F_VECTOR))
```
The `F_VECTOR` is a vector of `Polymake.Integer`s so this should be converted to `fmpz`, the numbers in the middle can grow very large.

The `f_vector` for `Polyhedron` could be adjusted as well but I cannot add a suggestion there, and probably `h_vector` and `g_vector` as well, but this could also be done in another PR.

File: docs/src/PolyhedralGeometry/cones.md [hunk 3]
```diff
@@ -63,12 +63,12 @@ f_vector(C::Cone)
 hilbert_basis(C::Cone)
```
Comment (benlorenz) [inline docs/src/PolyhedralGeometry/cones.md]:
```suggestion
hilbert_basis(C::Cone{fmpq})
```

File: src/PolyhedralGeometry/Cone/properties.jl [hunk 4]
```diff
@@ -393,13 +393,13 @@ julia> matrix(ZZ, hilbert_basis(C))
 """
 function hilbert_basis(C::Cone)
```
Comment (benlorenz) [inline src/PolyhedralGeometry/Cone/properties.jl]:
```suggestion
function hilbert_basis(C::Cone{fmpq})
```
Hilbert basis can only be computed for rational cones.

File: src/PolyhedralGeometry/LinearProgram.jl [hunk 5]
```diff
@@ -1,35 +1,26 @@
...
-
-"""
-struct LinearProgram
-   feasible_region::Polyhedron
-   polymake_lp::Polymake.BigObject
-   convention::Symbol
-end
-
-function LinearProgram(Q::Polyhedron, objective::AbstractVector; k = 0, convention = :max)
+function LinearProgram{T}(Q::Polyhedron, objective::AbstractVector; k = 0, convention = :max) where T<:scalar_types
```
Comment (benlorenz) [inline src/PolyhedralGeometry/LinearProgram.jl]:
```suggestion
function LinearProgram{T}(Q::Polyhedron{T}, objective::AbstractVector; k = 0, convention = :max) where T<:scalar_types
```
I don't think we want to allow mixing the types here (or doing an automatic conversion of the polyhedron).

File: src/PolyhedralGeometry/LinearProgram.jl [hunk 6]
```diff
@@ -1,35 +1,26 @@
...
-
-The linear program on the feasible set `P` (a Polyhedron) with
-respect to the function x ↦ dot(c,x)+k.
-
-"""
-struct LinearProgram
-   feasible_region::Polyhedron
-   polymake_lp::Polymake.BigObject
-   convention::Symbol
-end
```
Comment (benlorenz) [inline src/PolyhedralGeometry/LinearProgram.jl]: Can we move this back to this file? Having this in `iterators.jl` is kind of confusing.

Also why does it look like this file is stating with an `end`, but probably just github is being weird, maybe just for me?

File: src/PolyhedralGeometry/PolyhedralComplex/constructors.jl [hunk 7]
```diff
@@ -4,18 +4,11 @@
 ###############################################################################
 ###############################################################################
 
-struct PolyhedralComplex
-    pm_complex::Polymake.BigObject
-    function PolyhedralComplex(pm::Polymake.BigObject)
-        return new(pm)
-    end
-end
-
```
Comment (benlorenz) [inline src/PolyhedralGeometry/PolyhedralComplex/constructors.jl]: Can we move this back from iterators.jl?

File: src/PolyhedralGeometry/PolyhedralFan/constructors.jl [hunk 8]
```diff
@@ -1,53 +1,11 @@
...
-julia> iscomplete(PF)
-true
-"""
-struct PolyhedralFan <: _FanLikeType
-   pm_fan::Polymake.BigObject
-   function PolyhedralFan(pm::Polymake.BigObject)
-      return new(pm)
-   end
-end
-
```
Comment (benlorenz) [inline src/PolyhedralGeometry/PolyhedralFan/constructors.jl]: Move back?

File: src/PolyhedralGeometry/Polyhedron/properties.jl [hunk 9]
```diff
@@ -301,9 +301,9 @@ julia> facets(Halfspace, C)
 1: x₃ ≦ 1
 ```
 """
-facets(as::Type{T}, P::Polyhedron) where {R, S, T<:Union{AffineHalfspace, Pair{R, S}, Polyhedron}} = SubObjectIterator{as}(pm_object(P), _facet_polyhedron, nfacets(P))
+facets(as::Type{T}, P::Polyhedron) where {R, S<:scalar_types, T<:Union{AffineHalfspace{S}, Pair{R, S}, Polyhedron{S}}} = SubObjectIterator{as}(pm_object(P), _facet_polyhedron, nfacets(P))
```
Comment (benlorenz) [inline src/PolyhedralGeometry/Polyhedron/properties.jl]:
```suggestion
facets(as::Type{T}, P::Polyhedron{S}) where {R, S<:scalar_types, T<:Union{AffineHalfspace{S}, Pair{R, S}, Polyhedron{S}}} = SubObjectIterator{as}(pm_object(P), _facet_polyhedron, nfacets(P))
```

File: docs/src/PolyhedralGeometry/Polyhedra/auxiliary.md [hunk 10]
```diff
@@ -45,22 +45,22 @@ h_vector(P::Polyhedron)
 
 ```@docs
 all_triangulations
-boundary_lattice_points(P::Polyhedron)
+boundary_lattice_points(P::Polyhedron{fmpq})
 contains(P::Polyhedron, v::AbstractVector)
 ehrhart_polynomial(P::Polyhedron)
 ehrhart_polynomial(R::FmpqPolyRing, P::Polyhedron)
```
Comment (benlorenz) [inline docs/src/PolyhedralGeometry/Polyhedra/auxiliary.md]:
```suggestion
ehrhart_polynomial(P::Polyhedron{fmpq})
ehrhart_polynomial(R::FmpqPolyRing, P::Polyhedron{fmpq})
```

File: docs/src/PolyhedralGeometry/Polyhedra/auxiliary.md [hunk 11]
```diff
@@ -45,22 +45,22 @@ h_vector(P::Polyhedron)
...
+h_star_polynomial(R::FmpqPolyRing, P::Polyhedron{fmpq})
+interior_lattice_points(P::Polyhedron{fmpq})
+isnormal(P::Polyhedron{fmpq})
 issimple(P::Polyhedron)
-issmooth(P::Polyhedron)
+issmooth(P::Polyhedron{fmpq})
 is_very_ample(P::Polyhedron)
-lattice_points(P::Polyhedron)
+lattice_points(P::Polyhedron{fmpq})
 lattice_volume(P::Polyhedron)
```
Comment (benlorenz) [inline docs/src/PolyhedralGeometry/Polyhedra/auxiliary.md]:
```suggestion
lattice_volume(P::Polyhedron{fmpq})
```

File: docs/src/PolyhedralGeometry/Polyhedra/auxiliary.md [hunk 12]
```diff
@@ -45,22 +45,22 @@ h_vector(P::Polyhedron)
...
-interior_lattice_points(P::Polyhedron)
-isnormal(P::Polyhedron)
+h_star_polynomial(P::Polyhedron{fmpq})
+h_star_polynomial(R::FmpqPolyRing, P::Polyhedron{fmpq})
+interior_lattice_points(P::Polyhedron{fmpq})
+isnormal(P::Polyhedron{fmpq})
 issimple(P::Polyhedron)
-issmooth(P::Polyhedron)
+issmooth(P::Polyhedron{fmpq})
 is_very_ample(P::Polyhedron)
```
Comment (benlorenz) [inline docs/src/PolyhedralGeometry/Polyhedra/auxiliary.md]:
```suggestion
is_very_ample(P::Polyhedron{fmpq})
```

File: docs/src/PolyhedralGeometry/fans.md [hunk 13]
```diff
@@ -69,14 +69,14 @@ isregular(PF::PolyhedralFan)
 issimplicial(PF::PolyhedralFan)
 issmooth(PF::PolyhedralFan)
```
Comment (benlorenz) [inline docs/src/PolyhedralGeometry/fans.md]:
```suggestion
issmooth(PF::PolyhedralFan{fmpq})
```

File: src/PolyhedralGeometry/Polyhedron/standard_constructions.jl [hunk 14]
```diff
@@ -867,4 +868,6 @@ julia> volume(p)
 3
 ```
 """
-gelfand_tsetlin(lambda::AbstractVector) = Polyhedron(Polymake.polytope.gelfand_tsetlin(Vector{Rational}(lambda),projected=false))
+gelfand_tsetlin(::Type{T}, lambda::AbstractVector) where T<:scalar_types = Polyhedron{T}(Polymake.polytope.gelfand_tsetlin(Polymake.Vector{scalar_type_to_polymake[T]}(lambda), projected = false))
+
+gelfand_tsetlin(x...) = gelfand_tsetlin(fmpq, x...)
```
Comment (benlorenz) [inline src/PolyhedralGeometry/Polyhedron/standard_constructions.jl]:
```suggestion
gelfand_tsetlin(lambda::AbstractVector) = Polyhedron{fmpq}(Polymake.polytope.gelfand_tsetlin(Vector{Rational}(lambda), projected = false))
```

File: src/PolyhedralGeometry/SubdivisionOfPoints/constructors.jl [hunk 15]
```diff
@@ -4,13 +4,6 @@
 ###############################################################################
 ###############################################################################
 
-struct SubdivisionOfPoints
-   pm_subdivision::Polymake.BigObject
-   function SubdivisionOfPoints(pm::Polymake.BigObject)
-      return new(pm)
-   end
-end
-
```
Comment (benlorenz) [inline src/PolyhedralGeometry/SubdivisionOfPoints/constructors.jl]: move?

File: src/PolyhedralGeometry/Polyhedron/standard_constructions.jl [hunk 16]
```diff
@@ -840,13 +843,11 @@ julia> isfulldimensional(p)
...
 """
-project_full(P::Polyhedron) = Polyhedron(Polymake.polytope.project_full(pm_object(P)))
-
-
+project_full(P::Polyhedron{T}) where T<:scalar_types = Polyhedron{T}(Polymake.polytope.project_full(pm_object(P)))
 
 @doc Markdown.doc"""
 
-    gelfand_tsetlin(lambda::AbstractVector)
+    gelfand_tsetlin([::Type{T} = fmpq,] lambda::AbstractVector)
```
Comment (benlorenz) [inline src/PolyhedralGeometry/Polyhedron/standard_constructions.jl]:
```suggestion
    gelfand_tsetlin(lambda::AbstractVector)
```
Gelfand-Tsetlin polytope should stay with rational coordinates only, see https://github.com/oscar-system/Oscar.jl/pull/1022#issuecomment-1023041708.

File: src/PolyhedralGeometry/helpers.jl [hunk 17]
```diff
@@ -19,6 +19,43 @@ Base.convert(::Type{Polymake.Integer}, x::fmpz) = Polymake.Integer(BigInt(x))
...
+# Base.one(::Type{nf_scalar}) = fmpq(1)
+
+Base.convert(::Type{nf_scalar}, x::Number) = convert(fmpq, x)
+Base.convert(::Type{nf_scalar}, x::nf_elem) = x
+
+nf_scalar(x::Union{Number, nf_elem}) = convert(nf_scalar, x)
+
+function Base.convert(::Type{Polymake.QuadraticExtension{Polymake.Rational}}, x::nf_elem)
+    p = defining_polynomial(parent(x))
+    if (length(p) != 3 && coeff(p, 1) != 0) || sign(coeff(p, 2)) == sign(coeff(p, 0))
```
Comment (benlorenz) [inline src/PolyhedralGeometry/helpers.jl]:
```suggestion
    if length(p) != 3 || coeff(p, 1) != 0 || sign(coeff(p, 2)) == sign(coeff(p, 0))
```
Or maybe we can use something like this:
```suggestion
    isq = Hecke.isquadratic_type(parent(x))
    if isq[1] == true && isq[2] > 0
```

File: src/PolyhedralGeometry/helpers.jl [hunk 18]
```diff
@@ -19,6 +19,43 @@ Base.convert(::Type{Polymake.Integer}, x::fmpz) = Polymake.Integer(BigInt(x))
...
+        throw(ArgumentError("Conversion from nf_elem to QuadraticExtension{Rational} only defined for elements of number fields defined by a polynomial of the form 'ax^2 - b'."))
```
Comment (benlorenz) [inline src/PolyhedralGeometry/helpers.jl]:
```suggestion
        throw(ArgumentError("Conversion from nf_elem to QuadraticExtension{Rational} only defined for elements of real quadratic number fields defined by a polynomial of the form 'ax^2 - b'."))
```

File: src/ToricVarieties/NormalToricVarieties/attributes.jl [hunk 19]
```diff
@@ -687,7 +687,9 @@ A polyhedral fan in ambient dimension 2
 ```
 """
 @attr PolyhedralFan function fan(v::AbstractNormalToricVariety)
```
Comment (benlorenz) [inline src/ToricVarieties/NormalToricVarieties/attributes.jl]:
```suggestion
@attr PolyhedralFan{fmpq} function fan(v::AbstractNormalToricVariety)
```

File: src/ToricVarieties/NormalToricVarieties/attributes.jl [hunk 20]
```diff
@@ -687,7 +687,9 @@ A polyhedral fan in ambient dimension 2
...
-    return PolyhedralFan(pm_object(v))
+    return get_attribute!(v, :fan) do
+        return PolyhedralFan{fmpq}(pm_object(v))
+    end
```
Comment (benlorenz) [inline src/ToricVarieties/NormalToricVarieties/attributes.jl]:
```suggestion
    return PolyhedralFan{fmpq}(pm_object(v))
```
The `get_attribute` should not be necessary with the `@attr`, probably comes from a merge-error.

File: src/PolyhedralGeometry/iterators.jl [hunk 21]
```diff
@@ -224,9 +378,9 @@ function matrix_for_polymake(iter::SubObjectIterator; homogenized=false)
 end
 
 # primitive generators only for ray based iterators
-matrix(R::FlintIntegerRing, iter::SubObjectIterator{RayVector{Polymake.Rational}}) =
+matrix(R::FlintIntegerRing, iter::SubObjectIterator{<:RayVector}) =
```
Comment (benlorenz) [inline src/PolyhedralGeometry/iterators.jl]:
```suggestion
matrix(R::FlintIntegerRing, iter::SubObjectIterator{RayVector{fmpq}}) =
```
I don't think we want to call primitive for non-rational rays.

File: test/PolyhedralGeometry/Polyhedron.jl [hunk 22]
```diff
@@ -128,41 +150,61 @@
...
+            b1 = birkhoff(3)
+            b2 = birkhoff(3, even = true)
+            @test nvertices(pyramid(b1)) + 1 == nvertices(bipyramid(b1))
+            @test nvertices(b1) == nvertices(b2) * 2
+        end
+        P = gelfand_tsetlin(T, [3,2,1])
         p = project_full(P)
+        @test p isa Polyhedron{T}
         @test volume(P) == 0
         @test volume(p) == 1
```
Comment (benlorenz) [inline test/PolyhedralGeometry/Polyhedron.jl]:
```suggestion
            P = gelfand_tsetlin([3,2,1])
            p = project_full(P)
            @test p isa Polyhedron{T}
            @test volume(P) == 0
            @test volume(p) == 1
        end
```

**Review by benlorenz — COMMENTED** [review]
Thanks a lot for working on this, I think this looks very good already!

I added quite a few mostly small remarks.
A few general comments:
1. I added a few `{fmpq}` restrictions for some functions in the `docs` files but couldn't add suggestions in the corresponding src files because they don't appear in the diff.
2. The beginning of the iterators.jl files is quite confusing, can you clean this up a little bit and move some parts back to the original files? Maybe we need to extract some sections to a new file that is included earlier?
3. Can you add some testcase for a `QuadraticExtension` based standard construction like `dodecahedron` or `icosahedron`? (Or did I miss this? There are several testcases for polytopes using the new types but I didn't see anything that really uses the extra root...)

**Review by benlorenz — APPROVED** [review]
Thank you very much for all the work on this, this looks very good now and is good to go from my point of view (once the doctests succeed).

Note: There might still be various edge-cases where this can be improved further but we should get this merged to master as soon as possible to avoid any further conflicts and to allow more people to test and work with it.

**Review by benlorenz — APPROVED** [review]
doctests seem fine, printing the constraints looks a lot better now, thanks!

**Review by lkastner — APPROVED** [review]
